File: algorithms/Stack/python/155_minstack.py
from typing import List

class MinStack:

    def __init__(self):
        self.stack: List[int] = []
        self.min_stack: List[int] = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if(self.min_stack == [] or val <= self.min_stack[-1]):
            self.min_stack.append(val)
        

    def pop(self) -> None:
        
        if(self.min_stack[-1] == self.stack[-1]):
            self.min_stack.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

File: algorithms/Stack/cpp/155_minstack.cpp
#include <iostream>
#include <stack>
#include <limits.h>

using namespace std;

class MinStack {
private:
    stack<int> mainStack;
    stack<int> minStack;

public:
    MinStack() {
        
    }
    
    void push(int val) {
        mainStack.push(val);
        if(minStack.empty() || val <= minStack.top()) {
            minStack.push(val);
        }

    }
    
    void pop() {
        if(mainStack.top() == minStack.top()) {
            minStack.pop();
        }
        mainStack.pop();
    }
    
    int top() {
        return mainStack.top();
    }
    
    int getMin() {
        return minStack.top();
    }
};

int main() {
    MinStack* obj = new MinStack();
    obj->push(-2);
    obj->push(0);
    obj->push(-3);
    cout << obj->getMin() << endl; // Returns -3
    obj->pop();
    cout << obj->top() << endl;    // Returns 0
    cout << obj->getMin() << endl; // Returns -2
    delete obj;
    return 0;
}